File: solutions/swift/twelve-days/1/Sources/TwelveDays/TwelveDays.swift
let ordinals = [
  "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth",
  "eleventh", "twelfth",
]
let subjects = [
  "a Partridge", "two Turtle Doves", "three French Hens", "four Calling Birds",
  "five Gold Rings", "six Geese-a-Laying", "seven Swans-a-Swimming",
  "eight Maids-a-Milking", "nine Ladies Dancing", "ten Lords-a-Leaping",
  "eleven Pipers Piping", "twelve Drummers Drumming",
]
class TwelveDaysSong {
  private static func joinSubjects(_ n: Int) -> String {
    if n == 0 { return subjects[0] }
    if n == 1 { return subjects[1] + ", and " + subjects[0] }
    return subjects[n] + ", " + joinSubjects(n - 1)
  }
  private static func Verse(_ v: Int) -> String {
    "On the \(ordinals[v-1]) day of Christmas my true love gave to me: \(joinSubjects(v-1)) in a Pear Tree."
  }
  static func recite(start: Int, end: Int) -> String {
    (start...end)
      .map { Verse($0) }
      .joined(separator: "\n")
  }
}

File: solutions/zig/twelve-days/1/twelve_days.zig
const std = @import("std");

const days: [12][]const u8 = .{
    "first",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "eighth",
    "ninth",
    "tenth",
    "eleventh",
    "twelfth",
};

const gifts: [12][]const u8 = .{
    "twelve Drummers Drumming",
    "eleven Pipers Piping",
    "ten Lords-a-Leaping",
    "nine Ladies Dancing",
    "eight Maids-a-Milking",
    "seven Swans-a-Swimming",
    "six Geese-a-Laying",
    "five Gold Rings",
    "four Calling Birds",
    "three French Hens",
    "two Turtle Doves",
    "a Partridge in a Pear Tree.",
};

pub fn recite(buffer: []u8, start_verse: u32, end_verse: u32) []const u8 {
    var s = std.io.fixedBufferStream(buffer);
    for (start_verse - 1..end_verse) |i| {
        s.writer().print("On the {s} day of Christmas my true love gave to me: ", .{days[i]}) catch unreachable;
        for ((gifts.len - i - 1)..gifts.len) |j| {
            if (j > gifts.len - i - 1) {
                s.writer().writeAll(", ") catch unreachable;
                if (j == gifts.len - 1)
                    s.writer().writeAll("and ") catch unreachable;
            }
            s.writer().writeAll(gifts[j]) catch unreachable;
        }
        if (i < end_verse - 1)
            s.writer().writeAll("\n") catch unreachable;
    }
    return s.getWritten();
}
